pkg/stanza/adapter/receiver.go

```diff
@@ -132,15 +140,84 @@ func (r *receiver) consumerLoop(ctx context.Context) {
 				continue
 			}
 			obsrecvCtx := r.obsrecv.StartLogsOp(ctx)
-			cErr := r.consumer.ConsumeLogs(ctx, pLogs)
-			if cErr != nil {
-				r.logger.Error("ConsumeLogs() failed", zap.Error(cErr))
-			}
+			cErr := r.consumeLogs(ctx, pLogs)
 			r.obsrecv.EndLogsOp(obsrecvCtx, "stanza", pLogs.LogRecordCount(), cErr)
 		}
 	}
 }
 
+func (r *receiver) consumeLogs(ctx context.Context, logs plog.Logs) error {
+	if !r.retryConfig.Enabled {
+		err := r.consumer.ConsumeLogs(ctx, logs)
+		if err != nil {
+			r.logger.Error("ConsumeLogs() failed. " +
+				"Enable retry_on_failure to slow down reading logs and avoid dropping.", zap.Error(err))
+		}
+		return err
+	}
+
+	// Do not use NewExponentialBackOff since it calls Reset and the code here must
+	// call Reset after changing the InitialInterval (this saves an unnecessary call to Now).
+	expBackoff := backoff.ExponentialBackOff{
+		MaxElapsedTime:      r.retryConfig.MaxElapsedTime,
+		InitialInterval:     r.retryConfig.InitialInterval,
+		MaxInterval:         r.retryConfig.MaxInterval,
+		RandomizationFactor: backoff.DefaultRandomizationFactor,
+		Multiplier:          backoff.DefaultMultiplier,
+		Stop:                backoff.Stop,
+		Clock:               backoff.SystemClock,
+	}
+	expBackoff.Reset()
+
+	span := trace.SpanFromContext(ctx)
+	retryNum := int64(0)
+	for {
+		span.AddEvent(
+			"Sending logs.",
+			trace.WithAttributes(attribute.Int64("retry_num", retryNum)))
+
+		err := r.consumer.ConsumeLogs(ctx, logs)
+		if err == nil {
+			return nil
+		}
+
+		if consumererror.IsPermanent(err) {
+			r.logger.Error(
+				"ConsumeLogs() failed. The error is not retryable. Dropping data.",
+				zap.Error(err),
+				zap.Int("dropped_items", logs.LogRecordCount()),
```

Comment:
Shouldn't we extract number of dropped records from the `err` (as it should be `consumererror.Logs`)?

Reply:
Permanent errors don't have a number of dropped records yet. Will be added in https://github.com/open-telemetry/opentelemetry-collector/pull/7439

Reply:
I updated the logic to use partial retries tho.